cmd/nerdctl/login.go

```diff
@@ -240,14 +241,44 @@ func loginClientSide(ctx context.Context, cmd *cobra.Command, auth types.AuthCon
 	if len(regHosts) == 0 {
 		return "", fmt.Errorf("got empty []docker.RegistryHost for %q", host)
 	}
+	needFallbackToPlainHttp := false
 	for i, rh := range regHosts {
 		err = tryLoginWithRegHost(ctx, rh)
 		identityToken := fetchedRefreshTokens[rh.Host] // can be empty
 		if err == nil {
 			return identityToken, nil
 		}
+		if insecure && (errutil.IsErrHTTPResponseToHTTPSClient(err) || errutil.IsErrConnectionRefused(err)) {
+			needFallbackToPlainHttp = true
+			break
+		}
 		logrus.WithError(err).WithField("i", i).Error("failed to call tryLoginWithRegHost")
 	}
+	if needFallbackToPlainHttp {
+		dOpts = append(dOpts, dockerconfigresolver.WithPlainHTTP(true))
+		ho, err = dockerconfigresolver.NewHostOptions(ctx, host, dOpts...)
+		if err != nil {
+			return "", err
+		}
+		fetchedRefreshTokens := make(map[string]string) // key: req.URL.Host
+		// onFetchRefreshToken is called when tryLoginWithRegHost calls rh.Authorizer.Authorize()
+		onFetchRefreshToken := func(ctx context.Context, s string, req *http.Request) {
+			fetchedRefreshTokens[req.URL.Host] = s
+		}
+		ho.AuthorizerOpts = append(ho.AuthorizerOpts, docker.WithFetchRefreshToken(onFetchRefreshToken))
+		regHosts, err = dockerconfig.ConfigureHosts(ctx, *ho)(host)
+		if err != nil {
+			return "", err
+		}
+		for i, rh := range regHosts {
```

Comment:
no need to retry all `regHosts` you should fallback to http only in the registry that faced `(errutil.IsErrHTTPResponseToHTTPSClient(err) || errutil.IsErrConnectionRefused(err))`